java/source/infodynamics/measures/mixed/kraskov/MutualInfoCalculatorMultiVariateWithDiscreteKraskov.java

```diff
@@ -247,124 +461,130 @@ protected void computeNorms() {
+   *  (with PDFs computed using all of the previously supplied observation sets).</p>
+   *  
+   * <p>If the samples were supplied via a single call such as
+   * {@link #setObservations(double[][], int[])},
+   * then the return value is a single time-series of local
+   * channel measure values corresponding to these samples.</p>
+   * 
+   * <p>Otherwise where disjoint time-series observations were supplied using several 
+   *  calls such as {@link #addObservations(double[][], int[])}
+   *  then the local values for each disjoint observation set will be appended here
+   *  to create a single "time-series" return array.</p>
+   * 
+   * @return the "time-series" of local MIs in nats
+   * @throws Exception
+   */
+  public double[] computeLocalOfPreviousObservations() throws Exception {
+    return computeFromObservations(true);
+  }
+
+  /**
+   * This protected method handles the computation of either the average or
+   * local MI (over parts of the total observations).
+   * 
+   * <p>The method returns:<ol>
+   *  <li>for (returnLocals == false), an array of size 1,
+   *      containing the average MI </li>
+   *  <li>for local MIs (returnLocals == true), the array of local MI values</li>
+   *  </ol>
+   * 
+   * @param returnLocals whether to return an array or local values, or else
+   *  sums of these values
+   * @return either the average MI, or array of local MI value, in nats not bits
+   * @throws Exception
+   */
+	protected double[] computeFromObservations(boolean returnLocals) throws Exception {
+
+    int[] cumcount = new int[base];
+    Arrays.fill(cumcount, 0);
+
 
 		// Count the average number of points within eps_x and eps_y
 		double averageDiGammas = 0;
 		double avNx = 0;
 		double avNy = 0;
 		
 		double testSum = 0.0; // Used for debugging prints
-		
+    int N = totalObservations;
+    double[] locals = new double[N];
+
 		for (int t = 0; t < N; t++) {
-			// Compute eps_x and eps_y for this time step:
-			//  using x norms to all neighbours
-			//  (note that norm of point t to itself will be set to infinity).
-			// Then find the k closest neighbours in the same discrete bin
-			double eps_x = MatrixUtils.kthMinSubjectTo(xNorms[t], k, discreteData, discreteData[t]);			
-
-			// Count the number of points whose x distance is less
-			//  than or equal to eps_x (not including this point)
-			int n_x = 0;
-			for (int t2 = 0; t2 < N; t2++) {
-				if ((t2 != t) && (xNorms[t][t2] <= eps_x)) {
-					// xNorms has infinity for t == t2 anyway ...
-					n_x++;
-				}
-			}
-			// n_y is number of points in that bin, minus this point
-			int n_y = counts[discreteData[t]] - 1;
+
+      // Compute eps_x for this time step:
+      // find neighbours in the same discrete bin, then
+      // count points within eps_x in the whole dataset.
+      int b = discreteData[t];
+      double eps_x = kdTreeBins[b].findKNearestNeighbours(k, cumcount[b]++).poll().distance;

```

Comment:
We can't properly do the dynamic correlation exclusion time on the search in the full joint space (kdTreeBins[b]), since kdTreeBins[b] does not properly track the sample index for the point that it has in the full data set (i.e. for all b). Since we're not doing dynamic correlation exclusion on the full joint space, then I don't think we should do it in the reduced spaces either (for n_x from kdTreeJoint or for n_y), as we'll end up with strangeish numbers. One solution would be to somehow add tracking of these sample indices into kdTreeBins, but that would be rather complicated. Alternatively, it might be best if we simply remove the dynamic exclusion time property from this calculator altogether, so why don't we do that? If you want to leave the code here in case we come back to it later, you could remove the ability to set this parameter in setProperties, hard code dynCorrExclTime, and add a //TODO comment above the kdTreeBins[b] NN search to say that it would be good to add tracking of the proper sample indices into this data structure.

Reply:
I see. I actually did that sample index tracking in another calculator in the past, so I'll come back to this in the future and do it properly. For now I'm hard-coding dynCorrExclTime = 0 and removing it from setProperty. Changes in https://github.com/jlizier/jidt/pull/58/commits/ecda70f8e92859364c88b82153c6ab31cabdf91f